### PR TITLE
Refactor events

### LINKED ISSUE #609: Weird scrobbling behaviour (offset)
## Problem description

Ultrasonic emits some weird scrobbling heaviour for me. In particular, when playback is being started, the first track doesn't get scrobbled at all and subsequent ones are scrobbled with the timestamp of the previous one.

The moment the first song finishes playing, the second song is being scrobbled with the current timestamp. When the second song finishes playing, the third one is being scrobbled, and so on. When the last one finishes playing, nothing happens.

See also the screencast [here](https://files.catbox.moe/mkgj0c.mp4) (it was too big for GitHub).

### Steps to reproduce

- Open Ultrasonic
- Enable scrobbling
- Select some songs and press the "Play" icon

## System information

 * **Ultrasonic version**: *2.24.0 (F-Droid)*
 * **Android version**: *11 (LineageOS 18.1-20211010)*
 * **Device info**: *OnePlus 7 Pro (GM1917)*

## Additional notes

The now playing information is being updated correctly. I tested this with [gonic](https://github.com/sentriz/gonic) and [Navidrome](https://www.navidrome.org/), with the same results. When using Navidrome's built-in web interface (gonic doesn't have one), scrobbling works fine, so it doesn't seem to be a server issue.

### LINKED ISSUE #610: IllegalStateException
Stacktrace from G-Play:

```
java.lang.IllegalStateException: 
  at android.os.RemoteCallbackList.beginBroadcast (RemoteCallbackList.java:242)
  at android.support.v4.media.session.MediaSessionCompat$MediaSessionImplApi21.setPlaybackState (MediaSessionCompat.java:3915)
  at android.support.v4.media.session.MediaSessionCompat.setPlaybackState (MediaSessionCompat.java:814)
  at org.moire.ultrasonic.util.MediaSessionHandler.updateMediaSession (MediaSessionHandler.kt:252)
(....)
```

`MediaSessionCompat.java` calls `beginBroadcast()`. The exception comes from `RemoteCallbackList.java`. Here I find:
```
if (mBroadcastCount > 0) {
                throw new IllegalStateException(
                        "beginBroadcast() called while already in a broadcast");
            }
```

so it looks like calling `setPlaybackState` to frequently will cause an exception...

### PR DESCRIPTION
**Pull request by nitehu**
The Events we used were refactored to ReactiveX publishers/observables.

Originally I planned to add an EventBus, but it seems for Android the current best practice is to use Rx for this. It has numerous advantages, like very simple implementation of caching and throttling of the received events.
Instead of a central bus for all event types Rx can use separate publishers and observers for different types. I've kept all Rx parts in a separate singleton object for better decoupling. I'm not sure what are the established patterns for using Rx like this, let me know if it smells somewhere.

Fixes #610 
The exception in this issue was caused by accessing the MediaSessionHandler from concurrent threads. This is now fixed, as Rx processes all events on a single thread.

Fixes #609 
The issue was caused by that the time the player state change was sent, the currentPlaying track has changed. Now the player state change always contains the track it applies to.

Some minor problems have emerged as some concurrency disappeared from the code, I tried to hunt them down and fix them but this PR needs more testing.

### CODE REVIEW

**Comment by tzugen** [comment]
I just had a quick look at the code, it looks very promising

**Comment by nitehu** [comment]
I had some problems with the player state changes - the track and the state changes could run concurrently and sometimes they contained conflicting info. Now they are ordered and processed correctly.
These state changes are one of the most complex parts of Ultrasonic and they'd need some serious refactoring (maybe to a state machine).

File: ultrasonic/src/main/kotlin/org/moire/ultrasonic/service/Downloader.kt [hunk 1]
```diff
@@ -349,6 +353,20 @@ class Downloader(
...
+        while (iterator.hasNext()) {
+            val downloadFile = iterator.next()
+            if (!downloadFile.isCompleteFileAvailable) {
+                iterator.remove()
+                changedPlaylist = true
+            }
+        }
+        if (changedPlaylist) playlistUpdateRevision++
+    }
+
```
Comment (tzugen) [inline ultrasonic/src/main/kotlin/org/moire/ultrasonic/service/Downloader.kt:376]: Interesting :)
You added this function without using it, and I then started needing it in the next commit.

**Comment by tzugen** [comment]
I have put it on my phone and will test during the next days.

File: ultrasonic/src/main/kotlin/org/moire/ultrasonic/service/Downloader.kt [hunk 2]
```diff
@@ -46,7 +47,10 @@ class Downloader(
     private var wifiLock: WifiManager.WifiLock? = null
 
     var playlistUpdateRevision: Long = 0
-        private set
+        private set(value) {
+            field = value
+            RxBus.playlistPublisher.onNext(playlist)
+        }
```
Comment (tzugen) [inline ultrasonic/src/main/kotlin/org/moire/ultrasonic/service/Downloader.kt]: The playlist is now published with RX, while the observableDownloads is using LiveData.
I guess we should decide what is better and then use it for both.

File: ultrasonic/src/main/kotlin/org/moire/ultrasonic/service/MediaPlayerService.kt [hunk 3]
```diff
@@ -290,7 +285,6 @@ class MediaPlayerService : Service() {
             if (start) {
                 if (jukeboxMediaPlayer.isEnabled) {
                     jukeboxMediaPlayer.skip(index, 0)
-                    localMediaPlayer.setPlayerState(PlayerState.STARTED)
```
Comment (tzugen) [inline ultrasonic/src/main/kotlin/org/moire/ultrasonic/service/MediaPlayerService.kt:293]: Is this on purpose removed?

**Comment by tzugen** [comment]
I haven't encountered any problems so far :)

**Comment by tzugen** [comment]
Crash, kaboom...
````
2021-11-14 16:24:38.603 25319-25319/? E/SubsonicUncaughtExceptionHandler: Uncaught Exception! Android API level: 30
    Ultrasonic version name: 2.24.0
    Ultrasonic version code: 98
    
    
    io.reactivex.rxjava3.exceptions.OnErrorNotImplementedException: The exception was not handled due to missing onError handler in the subscribe() method call. Further reading: https://github.com/ReactiveX/RxJava/wiki/Error-Handling | java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.Object android.content.Context.getSystemService(java.lang.String)' on a null object reference
        at io.reactivex.rxjava3.internal.functions.Functions$OnErrorMissingConsumer.accept(Functions.java:718)
        at io.reactivex.rxjava3.internal.functions.Functions$OnErrorMissingConsumer.accept(Functions.java:715)
        at io.reactivex.rxjava3.internal.observers.LambdaObserver.onError(LambdaObserver.java:77)
        at io.reactivex.rxjava3.internal.observers.LambdaObserver.onNext(LambdaObserver.java:67)
        at io.reactivex.rxjava3.internal.util.NotificationLite.accept(NotificationLite.java:247)
        at io.reactivex.rxjava3.internal.operators.observable.ObservableReplay$BoundedReplayBuffer.replay(ObservableReplay.java:707)
        at io.reactivex.rxjava3.internal.operators.observable.ObservableReplay$ReplayObserver.replay(ObservableReplay.java:396)
        at io.reactivex.rxjava3.internal.operators.observable.ObservableReplay$ReplayObserver.onNext(ObservableReplay.java:361)
        at io.reactivex.rxjava3.internal.operators.observable.ObservableObserveOn$ObserveOnObserver.drainNormal(ObservableObserveOn.java:202)
        at io.reactivex.rxjava3.internal.operators.observable.ObservableObserveOn$ObserveOnObserver.run(ObservableObserveOn.java:256)
        at io.reactivex.rxjava3.android.schedulers.HandlerScheduler$ScheduledRunnable.run(HandlerScheduler.java:123)
        at android.os.Handler.handleCallback(Handler.java:938)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:223)
        at android.app.ActivityThread.main(ActivityThread.java:7664)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
     Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.Object android.content.Context.getSystemService(java.lang.String)' on a null object reference
        at android.view.LayoutInflater.from(LayoutInflater.java:282)
        at android.widget.ArrayAdapter.<init>(ArrayAdapter.java:213)
        at android.widget.ArrayAdapter.<init>(ArrayAdapter.java:207)
        at android.widget.ArrayAdapter.<init>(ArrayAdapter.java:193)
        at org.moire.ultrasonic.view.SongListAdapter.<init>(SongListAdapter.java:19)
        at org.moire.ultrasonic.fragment.PlayerFragment.onPlaylistChanged(PlayerFragment.kt:887)
        at org.moire.ultrasonic.fragment.PlayerFragment.onViewCreated$lambda-21(PlayerFragment.kt:426)
        at org.moire.ultrasonic.fragment.PlayerFragment.lambda$K0UNcb3OlWVMv1jQyGh3frF4aVI(Unknown Source:0)
        at org.moire.ultrasonic.fragment.-$$Lambda$PlayerFragment$K0UNcb3OlWVMv1jQyGh3frF4aVI.accept(Unknown Source:4)
        at io.reactivex.rxjava3.internal.observers.LambdaObserver.onNext(LambdaObserver.java:63)
        at io.reactivex.rxjava3.internal.util.NotificationLite.accept(NotificationLite.java:247) 
        at io.reactivex.rxjava3.internal.operators.observable.ObservableReplay$BoundedReplayBuffer.replay(ObservableReplay.java:707) 
        at io.reactivex.rxjava3.internal.operators.observable.ObservableReplay$ReplayObserver.replay(ObservableReplay.java:396) 
        at io.reactivex.rxjava3.internal.operators.observable.ObservableReplay$ReplayObserver.onNext(ObservableReplay.java:361) 
        at io.reactivex.rxjava3.internal.operators.observable.ObservableObserveOn$ObserveOnObserver.drainNormal(ObservableObserveOn.java:202) 
        at io.reactivex.rxjava3.internal.operators.observable.ObservableObserveOn$ObserveOnObserver.run(ObservableObserveOn.java:256) 
        at io.reactivex.rxjava3.android.schedulers.HandlerScheduler$ScheduledRunnable.run(HandlerScheduler.java:123) 
        at android.os.Handler.handleCallback(Handler.java:938) 
        at android.os.Handler.dispatchMessage(Handler.java:99) 
        at android.os.Looper.loop(Looper.java:223) 
        at android.app.ActivityThread.main(ActivityThread.java:7664) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947) 
2021-11-14 16:24:38.604 25319-25319/? I/SubsonicUncaughtExceptionHandler: Stack trace written to /storage/emulated/0/Android/data/org.moire.ultrasonic.debug/files/ultrasonic-stacktrace.txt
````

**Comment by nitehu** [comment]
Oh I didn't see this at the first glance...
The problem simply was that PlayerFragment should only subscribe for the duration while it is active (this is what the Disposable is used for in Rx)

**Comment by tzugen** [comment]
What do we need to do in order to finish this up?

**Comment by nitehu** [comment]
If the code seems to be OK for you, this can be merged. Maybe it needs a bit more testing, but we will test everything a few times anyway until the release (especially if we want to get anywhere with Api30 changes)

**Review by tzugen — APPROVED** [review]

**Pull request merged**